src/sleep/sleep.c

```diff
@@ -62,15 +62,31 @@ static int write_hibernate_location_info(void) {
                                        "Invalid hibernate type: %s", type);
 
         /* Only available in 4.17+ */
-        if (access("/sys/power/resume_offset", W_OK) < 0) {
+        if (access("/sys/power/resume_offset", R_OK) < 0) {
                 if (errno == ENOENT) {
                         log_debug("Kernel too old, can't configure resume offset, ignoring.");
                         return 0;
                 }
 
-                return log_debug_errno(errno, "/sys/power/resume_offset not writeable: %m");
+                return log_debug_errno(errno, "/sys/power/resume_offset not readable: %m");
         }
 
+        /* use resume_offset set by user if available */
+        r = read_one_line_file("/sys/power/resume_offset", &resume_offset);
+        if (r < 0)
+                log_warning_errno(r, "Error reading from /sys/power/resume_offset: %m");
+
+        r = safe_atou64(resume_offset, &offset);
```

Comment:
I don't understand what this patch is supposed to be doing. `offset` is read here, but it is overwritten below in line 104. The comment in TODO that is removed talks about the kernel command line, while `/sys/power/resume_offset` is read here. I'm confused.

Reply:
If `offset` is `> 0` we return on line 84 thus `offset` is only overwritten if we need to calculate an offset value. Would it be preferable to use a separate variable? That was the way I initially implemented it but seemed like `offset` could be safely reused so that was what I pushed.

I wrote a bit about the reason for for reading from `/sys/power/resume_offset` instead of the kernel cmdline in the bug [here](https://github.com/systemd/systemd/issues/11939#issuecomment-500185206) but in short, I don't think inspecting the cmdline values is the right solution here and that reading the most recent values from `/sys/power/resume_offset` is a better approach since A) the value set on the kernel cmdline will be already be present in `/sys/power/resume_offset` unless B) the value was changed after boot in which case the most recent value will be present in `/sys` and the cmdline would value could be invalid.

Similarly, I don't _think_ there is any need to check the `resume=` kernel cmdline or check `/sys/power/resume` explicitly because if a non-zero value is detected in `/sys/power/resume_offset`, we return before attempting to determine the `device` and write it to `/sys/power/resume` and instead rely on the current swap configuration present at the time of the hibernation request.

This approach implicitly honours whatever the user has configured and if the user has configured incorrectly, the hibernation will fail and log appropriately.

Happy to reconsider this if I'm thinking about it wrong or if there would be better ways to implement this.

Reply:
Thanks for the great feedback, @keszybz - this approach makes sense to me. I agree this will simplify the flow, put the checks in the right place, and make the logic easier to follow. I'll work on reworking this patch soon.